### ISSUE TITLE
Mask R-CNN fp16 train and test

### ISSUE THREAD

**Issue by XupingZHENG** [facebookresearch/Detectron#225]
I am considering the feasibility and difficulty of training and testing Mask R-CNN in fp16 mode. I list several points I have figured out. I am not sure whether I am on the right track.

1. Pretrained model loading
Either convert the pretrained ResNet or ResNeXt fp32 pkl model to fp16 mode. Or modifiy the source code, load the fp32 model and then convert fp32 weights to fp16 weights.

2. Weight initialization for head layers
Modify [fast_rcnn_heads.py](https://github.com/facebookresearch/Detectron/blob/master/lib/modeling/fast_rcnn_heads.py) and [mask_rcn_heads.py](https://github.com/facebookresearch/Detectron/blob/master/lib/modeling/mask_rcnn_heads.py), let WeightIntializer and BiasInitializer be pFP16Initializer. Enable tensor core and fp16 compute.
As in [resnet50_trainer.py](https://github.com/caffe2/caffe2/blob/master/caffe2/python/examples/resnet50_trainer.py).

3. Loss function
Convert fp16 back to fp32 before calculating the softmax loss and smooth l1 loss. Scaling may be needed to preserve accuracy, as pointed out by [the documentation](http://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html#caffe2).

4. RoiAlign
Implement fp16 version of RoiAlign layer.

Seems there is tons of work to do.

**Comment by xmyqsh** [comment]
@XupingZHENG 
Good instruction!

Three question:
1. After Pretrained model loading(convert fp32 weights to fp16 weights), how to set activations and gradients to fp16 in caffe2 easily?
2. I just found scale the loss up and down in loss scaling. It seems that there is no need to convert fp16 back to fp32 before calculating the loss.
3. FP32 Master Copy of Weights: How to holds a fp32 weights in caffe2 easily？
